File: src/main/java/com/kingmang/lazurite/parser/AST/Statements/FunctionDefineStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.libraries.Keyword;
import com.kingmang.lazurite.parser.AST.Arguments;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.UserDefinedFunction;
import lombok.AllArgsConstructor;

public record FunctionDefineStatement(String name, Arguments arguments, Statement body) implements Statement {

    @Override
    public void execute() {
        Keyword.put(name, new UserDefinedFunction(arguments, body));
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        if (body instanceof ReturnStatement) {
            return String.format("func %s%s = %s", name, arguments, ((ReturnStatement) body).expression);
        }
        return String.format("func %s%s %s", name, arguments, body);
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/FunctionDefineStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.libraries.Keyword
import com.kingmang.lazurite.parser.ast.Arguments
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.UserDefinedFunction

data class FunctionDefineStatement(
    val name: String,
    val arguments: Arguments,
    val body: Statement
) : Statement {
    override fun execute() {
        Keyword.put(this.name, UserDefinedFunction(this.arguments, this.body))
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        if (body is ReturnStatement)
            String.format("func %s%s = %s", this.name, this.arguments, this.body.expression)
        else String.format("func %s%s %s", this.name, this.arguments, this.body)
}
